src/scores/continuous/threshold_weighted_impl.py

```diff
@@ -0,0 +1,687 @@
+    """
+    The subderivative of `_phi_j_rect(a, b, x)` w.r.t. x.
+    """
+    return 4 * _g_j_rect(a, b, x)
+
+
+def _g_j_trap(a: EndpointType, b: EndpointType, c: EndpointType, d: EndpointType, x: xr.DataArray) -> xr.DataArray:
+    """
+    Returns values of a nondecreasing function g_j, where g_j is obtained by integrating
+    a trapezoidal weight function. The weight is 1 on the interval (b, c) and 0 outside
+    the interval (a,d). The formula is based on the third row of Table B1 from Taggart (2022).
+
+    Args:
+        a: left endpoint of interval where weight > 0.
+        b: left endpoint of the interval where weight = 1.
+        c: right endpoint of the interval where weight = 1.
+        d: right endpoint of interval where weight > 0.
+        x: points where g_j is to be evaluated.
+
+    Returns:
+        array of function values for the function g_j.
+
+    Note:
+        Requires a < b < c < d. This is tested in :py:func:`_auxiliary_funcs`.
+    """
+    # results correspond to each case in the third row of Table B1, Taggart (2022).
+    result0 = 0
+    result1 = (x - a) ** 2 / (2 * (b - a))
+    result2 = x - (b + a) / 2
+    result3 = -((d - x) ** 2) / (2 * (d - c)) + (d + c - a - b) / 2
+    result4 = (d + c - a - b) / 2
+    result = (
+        result1.where(x >= a, result0)
+        .where(x < b, result2)
+        .where(x < c, result3)
+        .where(x < d, result4)
+        .where(~np.isnan(x), np.nan)
+    )
+    return result
+
+
+def _phi_j_trap(a: EndpointType, b: EndpointType, c: EndpointType, d: EndpointType, x: xr.DataArray) -> xr.DataArray:
+    """
+    Returns values of a convex function phi_j, where phi_j is obtained by integrating
+    a trapezoidal weight function. The weight is 1 on the interval (b, c) and 0 outside
+    the interval (a,d). The formula is based on the fourth row of Table B1 from Taggart (2022).
+
+    Args:
+        a: left endpoint of interval where weight > 0.
+        b: left endpoint of the interval where weight = 1.
+        c: right endpoint of the interval where weight = 1.
+        d: right endpoint of interval where weight > 0.
+        x: points where phi_j is to be evaluated.
+
+    Returns:
+        array of function values for the function phi_j.
+
+    Note:
+        Requires a < b < c < d. This is tested in `_auxiliary_funcs`.
+    """
+    # results correspond to each case in the fourth row of Table B1, Taggart (2022).
+    result0 = 0
+    result1 = 2 * (x - a) ** 3 / (3 * (b - a))
+    result2 = 2 * x**2 - 2 * (a + b) * x + 2 * (b - a) ** 2 / 3 + 2 * a * b
+    result3 = (
+        2 * (d - x) ** 3 / (3 * (d - c))
+        + 2 * (d + c - a - b) * x
+        + 2 * ((b - a) ** 2 + 3 * a * b - (d - c) ** 2 - 3 * c * d) / 3
+    )
+    result4 = 2 * (d + c - a - b) * x + 2 * ((b - a) ** 2 + 3 * a * b - (d - c) ** 2 - 3 * c * d) / 3
+    result = (
+        result1.where(x >= a, result0)
+        .where(x < b, result2)
+        .where(x < c, result3)
```

Comment:
Just a suggestion, easier to debug by splitting this out into different lines rather than using chaining.

Reply:
Removed chaining